cpp/src/arrow/compute/row/grouper.h

```diff
@@ -30,6 +30,49 @@
 namespace arrow {
 namespace compute {
 
+/// \brief A segment of contiguous rows for grouping
+struct ARROW_EXPORT GroupingSegment {
+  int64_t offset;
+  int64_t length;
+  bool is_open;
+  bool extends;
+};
+
+inline bool operator==(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return segment1.offset == segment2.offset && segment1.length == segment2.length &&
+         segment1.is_open == segment2.is_open && segment1.extends == segment2.extends;
+}
+inline bool operator!=(const GroupingSegment& segment1, const GroupingSegment& segment2) {
+  return !(segment1 == segment2);
+}
+
+/// \brief Computes grouping segments for a batch. Each segment covers rows with identical
+/// values in the batch. The values in the batch are often selected as keys from a larger
+/// batch.
```

Comment:
```suggestion
/// \brief a helper class to divide a batch into segments of equal values
///
/// For example, given a batch with two rows:
///
/// A A
/// A A
/// A B
/// A B
/// A B
///
/// Then the batch could be divided into two segments.  The first would be rows 0 & 1
/// and the second would be rows 2, 3, & 4.
///
/// In addition, a segmenter keeps track of the last value seen.  This allows it to calculate
/// segments which span batches.  In our above example the last batch we emit would be
/// "open" which indicates that the segment may or may not extend into the next batch.
///
/// If the next call to the segmenter starts with `A B` then that segment would be marked with
/// `extends = true` so that the caller knows this is a continuation of the last open batch.
```

Reply:
Documented.